Source/ASDisplayNode.mm

```diff
@@ -1478,6 +1479,147 @@ - (void)recursivelySetNeedsDisplayAtScale:(CGFloat)contentsScale
   });
 }
 
+- (void)_layoutClipCornersIfNeeded
+{
+  ASDisplayNodeAssertMainThread();
+  if (_clipCornerLayers[0] == nil) {
+    return;
+  }
+  
+  CGSize boundsSize = self.bounds.size;
+  for (int idx = 0; idx < 4; idx++) {
+    BOOL isTop   = (idx == 0 || idx == 1);
+    BOOL isRight = (idx == 1 || idx == 2);
+    if (_clipCornerLayers[idx]) {
+      // Note the Core Animation coordinates are reversed for y; 0 is at the bottom.
+      _clipCornerLayers[idx].position = CGPointMake(isRight ? boundsSize.width : 0.0, isTop ? boundsSize.height : 0.0);
+      [_layer addSublayer:_clipCornerLayers[idx]];
+    }
+  }
+}
+
+- (void)_updateClipCornerLayerContentsWithRadius:(CGFloat)radius backgroundColor:(UIColor *)backgroundColor
+{
+  ASPerformBlockOnMainThread(^{
+    for (int idx = 0; idx < 4; idx++) {
+      // Layers are, in order: Top Left, Top Right, Bottom Right, Bottom Left.
+      // anchorPoint is Bottom Left at 0,0 and Top Right at 1,1.
+      BOOL isTop   = (idx == 0 || idx == 1);
+      BOOL isRight = (idx == 1 || idx == 2);
+      
+      CGSize size = CGSizeMake(radius + 1, radius + 1);
+      UIGraphicsBeginImageContextWithOptions(size, NO, self.contentsScaleForDisplay);
+      
+      CGContextRef ctx = UIGraphicsGetCurrentContext();
+      if (isRight == YES) {
+        CGContextTranslateCTM(ctx, -radius + 1, 0);
+      }
+      if (isTop == YES) {
+        CGContextTranslateCTM(ctx, 0, -radius + 1);
+      }
+      UIBezierPath *roundedRect = [UIBezierPath bezierPathWithRoundedRect:CGRectMake(0, 0, radius * 2, radius * 2) cornerRadius:radius];
+      [roundedRect setUsesEvenOddFillRule:YES];
+      [roundedRect appendPath:[UIBezierPath bezierPathWithRect:CGRectMake(-1, -1, radius * 2 + 1, radius * 2 + 1)]];
+      [backgroundColor setFill];
+      [roundedRect fill];
+      
+      // No lock needed, as _clipCornerLayers is only modified on the main thread.
+      CALayer *clipCornerLayer = _clipCornerLayers[idx];
+      clipCornerLayer.contents = (id)(UIGraphicsGetImageFromCurrentImageContext().CGImage);
+      clipCornerLayer.bounds = CGRectMake(0.0, 0.0, size.width, size.height);
+      clipCornerLayer.anchorPoint = CGPointMake(isRight ? 1.0 : 0.0, isTop ? 1.0 : 0.0);
+
+      UIGraphicsEndImageContext();
+    }
+    [self _layoutClipCornersIfNeeded];
+  });
+}
+
+- (void)_setClipCornerLayersVisible:(BOOL)visible
+{
+  ASPerformBlockOnMainThread(^{
+    ASDisplayNodeAssertMainThread();
+    if (visible) {
+      for (int idx = 0; idx < 4; idx++) {
+        if (_clipCornerLayers[idx] == nil) {
+          _clipCornerLayers[idx] = [[CALayer alloc] init];
+          _clipCornerLayers[idx].zPosition = 99999;
```

Comment:
I think following code will be better performance.
It reduce accessing to array.
```
CALayer *layer = [[CALayer alloc] init];
layer.zPosition = 99999;
layer.delegate = self;
_clipCornerLayers[idx] = layer;
```

Reply:
@appleguy 
Thanks! I understand.
Sorry, I may be misunderstanding about retain and release.
Does returning by subscript of NSArray not create a retain?